netbench-orchestrator/src/russula/network_utils.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::russula::{RussulaError, RussulaResult};
+use bytes::Bytes;
+use tokio::io::AsyncReadExt;
+use tokio::io::AsyncWriteExt;
+use tokio::net::TcpStream;
+use tracing::error;
+
+// Msg.len is represented as a u16
+const LEN_PREFIX_BYTES: usize = 2;
+
+/// Data format for messages exchanged with a peer.
+#[derive(Debug)]
+pub struct Msg {
```

Comment:
If feels like we're reinventing the wheel a bit here. Instead of dealing with the encoding ourselves, could we instead define a semantic message type (which represents the meaning of a message) and then just use serde for serialization/deserialization?

Reply:
Not sure I understand the suggestion here. Could you elaborate on what you mean.

Reply:
I already use [serde_json](https://github.com/aws/s2n-netbench/pull/29/files?diff=unified&w=1#diff-86f35409967dedc6617df9b08c5d65fbc0217b7be58c401245a052c22409f628R44-R49) to generate the payload.

The goal of Msg is to read exactly a Msg amount of bytes from the socket. In order to do that we need to prepend `len` to the payload and we need to only read `len` amount of bytes from the socket.

Reply:
> The goal of Msg is to read exactly a Msg amount of bytes from the socket

I agree that the Msg format accomplishes that, but it seems like the purpose of doing that is "we need a way to read json structs from the wire and deserialize them into Rust objects". My hope is that there is a way to do that without needing to write our own length-encoding scheme. My favorite code is no code at all 😄

Reply:
I agree with @jmayclin on this. I think you should be able to accomplish your goal of sending and receiving multiple messages through existing methods such as serde. For example, couldn't you just serialize a vec of messages and deserialize them on the other end, and then process them one by one? 

I think its nice you've gone to such lengths to define the message encoding, etc, but I could only really see that necessary in an absolutely performance critical implementations like `s2n-quic`. For Netbench we can tolerate the tiny bit of overhead that may be introduced by existing solutions, so I think its worth trying to pare things down to just the essential pieces that are unique to Netbench.

Reply:
> but I could only really see that necessary in an absolutely performance critical implementations

`Msg` doesn't add performance, it adds the ability to read 1 and only 1 payload at a time from the wire. That single payload can be json,txt,binary encoded. In the currently impl it is json encoded using the serde_json library.

> "we need a way to read json structs from the wire and deserialize them into Rust objects". My hope is that there is a way to do that without needing to write our own length-encoding scheme

I am not sure what a different solution looks like.

> For example, couldn't you just serialize a vec of messages and deserialize them on the other end, and then process them one by one?

What is 'message' here?

Reply:
> What is 'message' here?

Something higher level than what you have here, which looks like just some raw bytes. I would expect the message to be the actual instruction you want to send, probably an enum.